jax/lax/lax.py

```diff
@@ -3952,7 +3952,131 @@ def _scatter_dtype_rule(operand, scatter_indices, updates, **kwargs):
   _check_same_dtypes("scatter", False, operand.dtype, updates.dtype)
   return dtypes.canonicalize_dtype(operand.dtype)
 
-def _scatter_shape_rule(operand, scatter_indices, updates, **kwargs):
+def _scatter_shape_rule(operand, scatter_indices, updates, *, update_jaxpr,
+                        update_consts, dimension_numbers, indices_are_sorted,
+                        unique_indices):
+  """Validates the well-formedness of the ``dimension_numbers`` argument to
+  Scatter.
+
+  The code implements the checks based on the detailed operation semantics of
+  XLA's `Scatter <https://www.tensorflow.org/xla/operation_semantics#scatter>`_
+  operator and following the outline of the implementation of
+  ShapeInference::InferScatterShape in TensorFlow.
+  """
+  rank = lambda arr: len(arr.shape)
+
+  def _is_sorted(dims, name):
+    for i in range(1, len(dims)):
+      if dims[i] < dims[i - 1]:
+        raise TypeError(f"{name} in scatter op must be sorted; got {dims}")
+
+  def _sorted_dims_in_range(dims, rank, name):
+    if len(dims) == 0:
+      return
+    invalid_dim = None
+    if dims[0] < 0:
+      invalid_dim = dims[0]
+    elif dims[-1] >= rank:
+      invalid_dim = dims[-1]
+    if invalid_dim:
+      raise TypeError(f"Invalid {name} set in scatter op; valid range is "
+                      f"[0, {rank}); got: {invalid_dim}.")
+
+  def _no_duplicate_dims(dims, name):
```

Comment:
Do we need to check this? If it is sorted, then there cannot be duplicated dims, right?

Reply:
True – my sort is strict, which makes this check redundant. I wanted to mirror the code in XLA as close as possible, where these two checks happen in succession, but where sorting is not strict.
What do you think about changing sort to a non-strict sort and keeping `_no_duplicate_dims`? The rationale is that the errors thrown in JAX should then correspond exactly to the corresponding errors in XLA.

Reply:
(Note that `_no_duplicate_dims` is also used on `scatter_dims_to_dimension_dims`, which is not sorted!)